bpf/process/bpf_exit.c

```diff
@@ -1,17 +1,28 @@
 // SPDX-License-Identifier: GPL-2.0
 /* Copyright Authors of Cilium */
 
+#include "vmlinux.h"
 #include "bpf_exit.h"
+#include "bpf_tracing.h"
 
 char _license[] __attribute__((section("license"), used)) = "GPL";
 
-__attribute__((section("tracepoint/sys_exit"), used)) int
-event_exit(struct sched_execve_args *ctx)
+__attribute__((section("kprobe/__put_task_struct"), used)) int
+event_exit(struct pt_regs *ctx)
 {
-	__u64 pid_tgid;
+	struct task_struct *task =
+		(struct task_struct *)PT_REGS_PARM1_CORE(ctx);
+	__u32 pid, tgid;
 
-	pid_tgid = get_current_pid_tgid();
+	pid = BPF_CORE_READ(task, pid);
+	tgid = BPF_CORE_READ(task, tgid);
 
-	event_exit_send(ctx, pid_tgid);
+	/* We are only tracking group leaders so if tgid is not
+	 * the same as the pid then this is an untracked child
+	 * and we can skip the lookup/insert/delete cycle that
+	 * would otherwise occur.
+	 */
+	if (pid == tgid)
```

Comment:
Do we need this check here? Based on [this](https://elixir.bootlin.com/linux/v5.10/source/include/linux/sched/task.h#L112) it seems that it calls `__put_task_struct` only once per `task_struct`. 

Could we miss that event in the case where `pid` is not equal to `tgid` and this holds the last reference to the `task_struct`?

Reply:
there's task_struct for each thread and `__put_task_struct ` will be called for each

we want the exit event to be send just once and group leader is released when the
last child is released.. so having the (pid == tgid) check makes sure we send the
exit event just once for the leader when every other thread is dead

Reply:
> there's task_struct for each thread

Ok, missed that. Thanks